### PR TITLE
Increase metric name maximum length from 63 to 255 characters

### PR DESCRIPTION
**Pull request by trask**
Addresses part of #3422

## Changes

Increases metric name maximum length from 63 to 255 characters.

## Rationale

This limitation has been problematic in the Java and .NET communities (at least) where longer metric names are not uncommon:
* https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/6300
* https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/7448#issuecomment-1582846201
* https://github.com/open-telemetry/opentelemetry-java/issues/5656
* https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/740

As @aabmass mentions in https://github.com/open-telemetry/opentelemetry-specification/issues/3422#issue-1677012836

> Looks like this [the 63 character limitation] was introduced in #1557 by @reyang

but I didn't see any rationale in that PR for why that number was chosen.

## Is this a breaking change?

I don't think so, since it's a loosening (as opposed to strengthening) of the requirement, but you could also argue the alternative, since it's a "strengthening" of the requirement on telemetry consumers.

Alternatively, we could introduce a configurable setting for the metric name maximum length, but this would essentially add an (unbounded) strengthening of the requirement on telemetry consumers, so I'm not sure it's preferable.

### CODE REVIEW

**Review by reyang — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -23,6 +23,8 @@ release.
   ([#3613](https://github.com/open-telemetry/opentelemetry-specification/pull/3613))
 - Refine `MetricProvider.ForceFlush` and define `ForceFlush` for periodic exporting MetricReader.
   ([#3563](https://github.com/open-telemetry/opentelemetry-specification/pull/3563))
+- Increase metric name maximum length from 63 to 255 characters.
+  ([#9999](https://github.com/open-telemetry/opentelemetry-specification/pull/9999))
```
Comment (reyang) [inline CHANGELOG.md:27]:
```suggestion
  ([#3648](https://github.com/open-telemetry/opentelemetry-specification/pull/3648))
```

**Comment by reyang** [comment]
> ## Is this a breaking change?
> I don't think so, since it's a loosening (as opposed to strengthening) of the requirement, but you could also argue the alternative, since it's a "strengthening" of the requirement on telemetry consumers.
> 
> Alternatively, we could introduce a configurable setting for the metric name maximum length, but this would essentially add an (unbounded) strengthening of the requirement on telemetry consumers, so I'm not sure it's preferable.

+1

**Review by mateuszrzeszutek — APPROVED** [review]

**Review by jack-berg — APPROVED** [review]

**Review by jsuereth — APPROVED** [review]

**Comment by joaopgrassi** [comment]
I did some looking on what most o11y vendors tell on their public pages about metric name/key size just as an FYI

- Dynatrace: 250 [source](https://www.dynatrace.com/support/help/extend-dynatrace/opentelemetry/getting-started/metrics/limitations#limits)
- Lightstep: 100 (I think?) [source](https://docs.lightstep.com/docs/otel-metrics-concepts#frequently-asked-questions)
- Datadog: 200 [source](https://docs.datadoghq.com/metrics/custom_metrics/#naming-custom-metrics)
- New Relic: 255 (I think?) [source](https://docs.newrelic.com/docs/data-apis/ingest-apis/metric-api/metric-api-limits-restricted-attributes/)

I think it should be fine since we are increasing the limit. If a back-end that accepted less than 255 truncated the metric name to fit their limit, would still continuing truncating as usual.

**Comment by reyang** [comment]
@trask would you update the changelog by moving the entry from 1.24.0 to unreleased?

**Comment by nayaniabhishek** [comment]
This is still not fixed for Java: https://github.com/open-telemetry/opentelemetry-java/blob/b3376a25d06bca773cfd16f600b660a0badf37a8/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeter.java#L49

**Comment by trask** [comment]
hi @nayaniabhishek, here's the Java PR: https://github.com/open-telemetry/opentelemetry-java/pull/5697 (I'm also looking forward to this change)

but please note that it won't be merged until there has been a specification release that includes this spec change (see https://github.com/open-telemetry/opentelemetry-specification/issues/2728#issuecomment-1247296265)

**Comment by kalkibhagavanmeda** [comment]
Hi team
Any expectations like when will be releasing latest version of opentelemetry-api with increased metrics name length limitation

**Comment by trask** [comment]
hi @kalkibhagavanmeda, I'd suggest asking this in the language repository for the SDK you are using, e.g. this is implemented in the latest released Java SDK version (https://github.com/open-telemetry/opentelemetry-java/pull/5697)

**Pull request merged**